src/DotNetEnv/Parsers.cs

```diff
@@ -173,17 +173,23 @@ private static string ToUtf32Char (string hex)
                 .Or(OctalChar)
                 .Or(EscapedChar);
 
+        internal static readonly Parser<IEnumerable<char>> InlineCommentBeginOrControl =
+            Parse.Regex("[ \t]#").Or(Parse.Char(c => char.IsControl(c) && c != '\t', "controls except tab").Once());
+
         // unquoted values can have interpolated variables,
         // but only inline whitespace -- until a comment,
         // and no escaped chars, nor byte code chars
-        // FIXME: would be nice to solve multi word with comment case with parser directly (no split + trim)
         internal static readonly Parser<ValueCalculator> UnquotedValue =
-            InterpolatedValue.Or(
-                NotControlNorWhitespace("'\"$")
-                .Or(InlineWhitespaceChars.AtLeastOnce().Text())
-                .AtLeastOnce()
-                .Select(strs => new ValueActual(strs))
-            ).Many().Select(vs => new ValueCalculator(vs).Split("[ \t]#").Trim());
+            InterpolatedValue
+                .Or(Parse.CharExcept('$').Except(InlineCommentBeginOrControl)
```

Comment:
This is the bit that I feel makes your current approach more complicated than mine in #87 -- you attempted to retain too much of the old (undesired) behavior of the previous implementation, which is that the unquoted value continues to parse comments itself, and then discard them, rather than leaving that responsibility to the Assignment parser which discards comments for single + double quoted comments already.

My PR changes it so that the unquoted value parser *does not read comments at all* so the Assignment parser continues to do it's job the way it's supposed to, and discards comments there, consistent with the other kinds of values.

Your PR continues to parse comments as a part of unquoted value, which is, at the core, the behavior that the FIXME comment was aiming at stopping (as identified as being about the regex / splitting and trimming).

Reply:
Changed an adapted some of your code, I'm closing this, if there are additional things just reopen or add new discussion ;)